codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -174,56 +76,19 @@ public T get(T name, T defaultValue) {
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T getAndRemove(T name) {
-        checkNotNull(name, "name");
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        HeaderEntry e = entries.get(i);
-        if (e == null) {
-            return null;
-        }
-
-        T value = null;
-        for (;;) {
-            if (e.hash == h && keyComparator.compare(e.name, name) == 0) {
-                if (value == null) {
-                    value = e.value;
-                }
-                e.remove();
-                HeaderEntry next = e.next;
-                if (next != null) {
-                    entries.put(i, next);
-                    e = next;
-                } else {
-                    entries.remove(i);
-                    tailEntries.remove(i);
-                    return value;
-                }
-            } else {
-                break;
-            }
+        Object value = map.remove(name);
+        if (isList(value)) {
+            List<T> value0 = (List<T>) value;
+            size -= value0.size();
+            return value0.get(0);
         }
-
-        for (;;) {
-            HeaderEntry next = e.next;
-            if (next == null) {
-                break;
-            }
-            if (next.hash == h && keyComparator.compare(e.name, name) == 0) {
-                if (value == null) {
-                    value = next.value;
-                }
-                e.next = next.next;
-                if (e.next == null) {
-                    tailEntries.put(i, e);
-                }
-                next.remove();
-            } else {
-                e = next;
-            }
+        if (value != null) {
```

Comment:
Consider replacing this with a `if (value == null) { return; }` after the `Object value = map.remove(name);` line to clarify that null is not expected/allowed.